File: pe-core/src/main/java/org/saar/core/common/flatreflected/FlatReflectedModel.java
package org.saar.core.common.flatreflected;

import org.joml.Vector3f;
import org.saar.core.mesh.Model;
import org.saar.maths.transform.SimpleTransform;

public class FlatReflectedModel implements Model {

    private final FlatReflectedMesh mesh;
    private final Vector3f normal;

    private final SimpleTransform transform = new SimpleTransform();

    public FlatReflectedModel(FlatReflectedMesh mesh, Vector3f normal) {
        this.mesh = mesh;
        this.normal = normal;
    }

    public Vector3f getNormal() {
        return this.normal;
    }

    public SimpleTransform getTransform() {
        return this.transform;
    }

    @Override
    public void draw() {
        this.mesh.draw();
    }

    @Override
    public void delete() {
        this.mesh.delete();
    }
}

File: pe-core/src/main/java/org/saar/core/common/flatreflected/FlatReflectedModel.kt
package org.saar.core.common.flatreflected

import org.joml.Vector3f
import org.saar.core.mesh.Mesh
import org.saar.core.mesh.Model
import org.saar.lwjgl.opengl.texture.ReadOnlyTexture
import org.saar.lwjgl.opengl.texture.Texture2D
import org.saar.maths.transform.SimpleTransform

class FlatReflectedModel(
    override val mesh: Mesh,
    val normal: Vector3f,
    val transform: SimpleTransform,
) : Model {

    var reflectionMap: ReadOnlyTexture = Texture2D.NULL

    constructor(mesh: Mesh, normal: Vector3f) : this(mesh, normal, SimpleTransform())

    override fun delete() {
        super.delete()
        this.reflectionMap.delete()
    }
}